lib/api/apiUtils/object/versioning.js

```diff
@@ -370,6 +402,69 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD,
     return callback(null, options);
 }
 
+/** overwritingVersioning - return versioning information for S3 to handle
+ * storing version metadata with a specific version id.
+ * @param {string} bucketName - name of the bucket.
+ * @param {string} putVersionId - version id from x-scal-s3-version-id header
+ * @param {object} objMD - obj metadata of the master version
+ * @param {object} metadataStoreParams - custom built object containing resource details.
+ * @param {RequestLogger} log - logger instance
+ * @param {function} cb - cb(err, options)
+ * options.versionId - specific versionId to overwrite in metadata
+ *  ('' overwrites the master version)
+ * options.versioning - (true/undefined) metadata instruction to create new ver
+ * options.isNull - (true/undefined) whether new version is null or not
+ * options.nullVersionId - if storing a null version in version history, the
+ *  version id of the null version
+ * @return {undefined}
+ */
+function overwritingVersioning(bucketName, putVersionId, objMD, metadataStoreParams, log, cb) {
+    const { err, decodedVid: decodedPutVersionId, isNull } = _decodePutVersionId(putVersionId);
+    if (err) {
+        log.error('invalid x-scal-s3-version-id header value', { error: err, putVersionId });
+        return cb(err);
+    }
+
+    /* eslint-disable no-param-reassign */
+    // CASE 1: overwriting the master null version or,
+    //         overwriting a non-versioned object.
+    if (isNull && (objMD.isNull || !objMD.versionId)) {
+        metadataStoreParams.creationTime = objMD['creation-time'];
+        metadataStoreParams.lastModifiedDate = objMD['last-modified'];
+        // masterVersionId is used instead of generating a new master version id.
+        const options = { masterVersionId: objMD.versionId || '' };
+        // NOTE: if object is not versioned we do not store the isNull property.
+        if (objMD.versionId) {
+            options.isNull = true;
+        }
+        return process.nextTick(() => cb(null, options));
+    }
+
+    // CASE 2: overwriting a version id.
+    const options = {
+        versionId: decodedPutVersionId,
+    };
+
+    // NOTE: We need to retrieve the specific object version ("versionMD") to ensure
+    //       that the rest of the original object properties remain intact.
+    // NOTE: "objMD" holds the master version properties,
+    return metadata.getObjectMD(bucketName, metadataStoreParams.objectKey, options, log,
```

Comment:
Currently looking for a clean way to leverage objMD and avoid having to call object metadata twice.